jep/234/README.adoc

```diff
@@ -0,0 +1,274 @@
+
+[source,java]
+----
+package jenkins.views;
+
+import hudson.Extension;
+
+@Extension(ordinal = Integer.MIN_VALUE)
+public class JenkinsHeader extends Header {
+
+    @Override
+    public boolean isHeaderEnabled() {
+        return true;
+    }
+    [...]
+}
+----
+
+* As mentioned before, method `get()` from `JenkinsHeader` will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value)
+
+[source,java]
+----
+[...]
+@Restricted(NoExternalUse.class)
+@CheckForNull
+public static Header get() {
+    List<Header> headers = ExtensionList.lookup(Header.class).stream()
+                .filter(header -> header.isHeaderEnabled())
+                .collect(Collectors.toList());
+    if (headers.size() > 0) {
+        if (headers.size() > 1) {
+            LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
+        } else {
+            return headers.get(0);
+        }
+    }
+    return new JenkinsHeader();
+}
+----
+
+* Once we launch Jenkins with the proposed changes on the core, we will obtain the expected/current header working without any issue
+
+=== Custom UI plugin
+
+* Create a new plugin following the usual procedure
+* Provide an implementation of the custom Header (e.g: `src/main/java/org/jenkinsci/plugins/custom/header/CustomHeader.java`)
+
+[source,java]
+----
+[...]
+@Extension(ordinal = 100)
+public class CustomHeader extends Header {
+
+    @Override
+    public boolean isHeaderEnabled() {
+        // Disable/enable the header based on an ENV var and/or system property
+        boolean isDisabled = System.getProperty(CustomHeader.class.getName() + ".disable") != null ?
+                "true".equalsIgnoreCase(System.getProperty(CustomHeader.class.getName() + ".disable")) :
+                "true".equalsIgnoreCase(System.getenv("CUSTOM_HEADER_DISABLE"));
+        return !isDisabled;
+    }
+}
+----
+
+* Provide a method in the custom header to retrieve the label which will be with the username. Current code is just an example, but the label could be obtained from the https://javadoc.jenkins.io/jenkins/model/GlobalConfiguration.html[GlobalConfiguration].
+
+[source,java]
+----
+    public static String getHeaderLabel(){
+        // This label content could be retrieved programatically. Not coded in aims of simplicity.
+        return "Hello World!";
+    }
+----
+
+* Provide the jelly file to override the `headerContent`. For that purpose, use the common location convention. For the previous example: `src/main/resources/org/jenkinsci/plugins/custom/header/CustomHeader/`. Retrieve the customizable label to be rendered with the username on the `headerContent` file.
+
+```xml
+<j:invokeStatic var="label" className="org.jenkinsci.plugins.custom.header.CustomHeader" method="getHeaderLabel"/>
```

Comment:
Last I checked `invokeStatic` does not work on plugin classes. Did you actually test this? (In `RealJenkinsRule` or using `java -jar jenkins.war`, _not_ using `JenkinsRule` or `mvn hpi:run`!)

I think you rather want to ensure that `it` is correctly bound when rendering Jelly.

Reply:
Tested using `java -jar jenkins.war` and working without issues